src/helpers/qa.ts

```diff
@@ -0,0 +1,659 @@
+      dedupChains.push(chain);
+      seen.push(stringifyChain(chain));
+    }
+  });
+  return dedupChains;
+}
+
+function stringifyChain(chain: Chain): string {
+  return `${chain.chainType}_${chain.chainId}`;
+}
+
+type QaOptions = {
+  maker: string;
+  chain: number;
+  chain_type: string;
+  quote_chain?: number;
+  quote_chain_type: string;
+  check_all_xchain: boolean;
+  base_token?: string;
+  quote_token?: string;
+  env: string;
+  num_requests: number;
+  delay_ms: number;
+};
+
+export async function runQa(
+  argv: QaOptions,
+  name: string,
+  authKey: string
+): Promise<number> {
+  const maker = argv.maker;
+  const chainType = argv.chain_type as ChainType;
+  const chainId = argv.chain as ChainId;
+  const chain: Chain = { chainType, chainId };
+  const env = argv.env as Environment;
+  validateMakerName(maker);
+  validateChain(chain);
+  validateEnvironment(env);
+
+  const crossChainCheck = argv.check_all_xchain;
+
+  if (crossChainCheck && argv.quote_chain) {
+    throw new Error(
+      `Can't specify a quote_chain and check_all_xchain simultaneously`
+    );
+  }
+
+  const maybeQuoteChain = argv.quote_chain
+    ? {
+        chainType: argv.quote_chain_type as ChainType,
+        chainId: argv.quote_chain as ChainId,
+      }
+    : undefined;
+  if (maybeQuoteChain) {
+    validateChain(maybeQuoteChain);
+  }
+
+  const evmQaAddress = process.env.QA_TAKER_ADDRESS?.toLowerCase();
+  if (evmQaAddress) {
+    validateEvmAddress(evmQaAddress);
+  }
+  const solanaQaAddress = process.env.QA_TAKER_ADDRESS_SOLANA;
+  if (solanaQaAddress) {
+    validateSolanaAddress(solanaQaAddress);
+  }
+
+  const hashflow = new HashflowApi('taker', name, authKey, env);
+
+  const numRequests = argv.num_requests;
+  const delayMs = argv.delay_ms;
+  const delayMsStr =
+    delayMs > 0 ? ` using a delay of ${delayMs}ms between RFQs` : '';
+
+  const pairProvided = argv.base_token && argv.quote_token;
+
+  process.stdout.write(
+    `QA testing maker '${maker}' against ${env} on chain ${chain.chainType}:${
+      chain.chainId
+    } ${
+      pairProvided ? ` for ${argv.base_token}-${argv.quote_token}` : ''
+    } with ${numRequests} requests/pair${delayMsStr}.\n\n`
+  );
+
+  process.stdout.write('Finding active makers ... ');
+  const makers: string[] = [];
+  try {
+    const chainMakers = await hashflow.getMarketMakers(chain, undefined, maker);
+    for (const externalMaker of chainMakers) {
+      const makerPrefix = externalMaker.split('_')[0];
+      if (makerPrefix === maker) {
+        makers.push(externalMaker);
+      }
+    }
+
+    if (!makers.length) {
+      process.stdout.write(
+        `failed! No maker available ${JSON.stringify(chainMakers)}\n`
+      );
+      return 1;
+    }
+  } catch (err) {
+    process.stdout.write(`failed!  ${(err as Error).toString()}\n`);
+    return 1;
+  }
+
+  const makersString = makers.length > 1 ? JSON.stringify(makers) : makers[0];
+  process.stdout.write(`done.  ${makersString}\n`);
```

Comment:
These process.stdout.write calls may seriously pollute the logs/output. How about we pass in a function here so that the script can pass `process.stdout.write` and API callers / hermes can pass in a function that collects the output. (or just ignores it)

Reply:
fixed